src/main/networking/vtunnel.ts

```diff
@@ -31,24 +43,70 @@ export class vtunnel {
     },
   });
 
-  async generateVtunnelConfig() {
+  // eslint-disable-next-line no-useless-constructor
+  private constructor() {}
+
+  static getInstance(): Vtunnel {
+    return Vtunnel._instance;
+  }
+
+  private async generateConfig() {
     const conf = {
-      tunnel: [
-        {
-          'handshake-port':          credServerHandshakePort,
-          'vsock-host-port':         credServerVsockPort,
-          'peer-address':            localHost,
-          'peer-port':               credServerPeerPort,
-          'upstream-server-address': `${ localHost }:${ credServerPort }`,
-        }
-      ]
+      tunnel: this._vtunnelConfig.map(c => ({
+        'handshake-port':          c.handshakePort,
+        'vsock-host-port':         c.vsockHostPort,
+        'peer-address':            c.peerAddress,
+        'peer-port':               c.peerPort,
+        'upstream-server-address': c.upstreamServerAddress,
+      }))
     };
+
     const configYaml = yaml.stringify(conf);
 
     await fs.promises.writeFile(getVtunnelConfigPath(), configYaml, 'utf8');
   }
+
+  /**
+   * addTunnel adds a new configuration to an existing list of configs.
+   * @param config
+   */
+  addTunnel(config: VtunnelConfig) {
+    this._vtunnelConfig.push(config);
+  }
+
+  /**
+   * start generates the final configuration yaml file and starts the
+   * Vtunnel Host process.
+   */
+  async start() {
+    try {
+      await this.generateConfig();
+    } catch (error) {
+      console.log(`Failed to generate vtunnel configuration.`);
```

Comment:
Actually, if `generateConfig` threw, shouldn't we avoid calling `this.vsockProxy.start()`? It's not like it's going to work…